pyomo/core/base/expression.py

```diff
@@ -361,29 +361,12 @@ def store_values(self, new_values):
         for index, new_value in iteritems(new_values):
             self._data[index].set_value(new_value)
 
-    def _default(self, index):
-        raise KeyError(str(index))
-
-    def __setitem__(self, ndx, val):
-        #
-        # Get the expression data object
-        #
-        exprdata = None
-        if ndx in self._data:
-            exprdata = self._data[ndx]
-        else:
-            _ndx = normalize_index(ndx)
-            if _ndx in self._data:
-                exprdata = self._data[_ndx]
-        if exprdata is None:
-            raise KeyError(
-                "Cannot set the value of Expression '%s' with "
-                "invalid index '%s'"
-                % (self.name, str(ndx)))
-        #
-        # Set the value
-        #
-        exprdata.set_value(val)
+    def _getitem_when_not_present(self, index):
```

Comment:
FWIW, I don't think it makes sense to implicitly create a named expression.  Vars are a special case in many ways.  If we think we need this, we should have motivating examples.

Reply:
I agree.  I put it in because that is how things currently work.  However, I thing currently work that way because in the old system we did not distinguish between being able to implicitly _create_ a `_ComponentData` when it was not already present and being able to implicitly _get_ a `_ComponentData` when it was not already present.  For Expression components, I believe that the former is fine and the latter makes no sense.  I am hoping that @ghackebeil will comment.  If he is OK with it, I would like to remove `Expression._getitem_when_not_present()`.

Reply:
I think you'll need to bug him for input if you want it before the holidays.